aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -409,10 +400,270 @@ impl AptosVM {
+                serialize_values(&vec![
+                    MoveValue::Address(txn_payload.multisig_address),
+                    MoveValue::vector_u8(provided_payload),
+                ]),
+                gas_meter,
+            )?
+            .return_values
+            .into_iter()
+            .map(|(bytes, _ty)| bytes)
+            .collect::<Vec<_>>();
+        let payload_bytes = payload_bytes
+            .first()
+            // We expect the payload to either exists on chain or be passed along with the
+            // transaction.
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .finish(Location::Undefined)
+            })?;
+        // We have to deserialize twice as the first time returns the actual return type of the
+        // function, which is vec<u8>. The second time deserializes it into the correct
+        // EntryFunction payload type.
+        // If either deserialization fails for some reason, that means the user provided incorrect
+        // payload data either during transaction creation or execution.
+        let deserialization_error = PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT)
+            .finish(Location::Undefined);
+        let payload_bytes =
+            bcs::from_bytes::<Vec<u8>>(payload_bytes).map_err(|_| deserialization_error.clone())?;
+        let payload = bcs::from_bytes::<MultisigTransactionPayload>(&payload_bytes)
+            .map_err(|_| deserialization_error)?;
+
+        // Step 2: Execute the target payload. Transaction failure here is tolerated. In case of any
+        // failures, we'll discard the session and start a new one. This ensures that any data
+        // changes are not persisted.
+        // The multisig transaction would still be considered executed even if execution fails.
+        let execution_result = match payload {
+            MultisigTransactionPayload::EntryFunction(entry_function) => self
+                .execute_multisig_entry_function(
+                    &mut session,
+                    gas_meter,
+                    txn_payload.multisig_address,
+                    &entry_function,
+                    new_published_modules_loaded,
+                ),
+        };
+
+        // Step 3: Call post transaction cleanup function in multisig account module with the result
+        // from Step 2.
+        // Note that we don't charge execution or writeset gas for cleanup routines. This is
+        // consistent with the high-level success/failure cleanup routines for user transactions.
+        let cleanup_args = serialize_values(&vec![
+            MoveValue::Address(txn_data.sender),
+            MoveValue::Address(txn_payload.multisig_address),
+            MoveValue::vector_u8(payload_bytes),
+        ]);
+        let final_change_set_ext = if let Err(execution_error) = execution_result {
+            self.failure_multisig_payload_cleanup(
```

Comment:
this doesn't seem cleaning up the pending module published with `new_published_modules_loaded` variable

Reply:
new_published_modules_loaded is passed down to execute_multisig_entry_function on line 509, where it would be marked if there is a module publish. The cache is then marked as invalidated in error handling in execute_user_transaction which is the one calling execute_multisig_transaction

Reply:
Good catch! The invalidation only happens upon error handling in execute_user_transaction, which would not trigger because this would not return an error. Let me add logic here to handle the invalidation then. It's unfortunately duplicate

Reply:
Updated